gsy_framework/sim_results/carbon_emissions/results.py

```diff
@@ -0,0 +1,346 @@
+            df_carbon_ratio = pd.DataFrame(index=full_range)
+            df_carbon_ratio["Ratio (gCO2eq/kWh)"] = df_file[
+                "Carbon intensity of electricity - gCO2/kWh"
+            ]
+
+        # fill df_carbon_ratio in case of missing hours using the last value
+        df_carbon_ratio = df_carbon_ratio[["Ratio (gCO2eq/kWh)"]]
+        df_carbon_ratio = df_carbon_ratio.reindex(
+            pd.date_range(start=start, end=end, freq="H", tz="UTC"), method="pad"
+        )
+        return df_carbon_ratio
+
+    def calculate_from_carbon_ratio_and_imported_energy(
+        self, df_carbon_ratio: pd.DataFrame, df_imported_energy: pd.DataFrame
+    ):
+        """Calculate the carbon emissions from a dataframe with carbon ratios
+        and another with imported/exported energy."""
+        if not {"Ratio (gCO2eq/kWh)"}.issubset(df_carbon_ratio.columns):
+            raise ValueError(
+                "df_carbon_ratio must contain columns: simulation_time, Ratio (gCO2eq/kWh)"
+            )
+        if not {
+            "simulation_time",
+            "area_uuid",
+            "imported_from_grid",
+            "imported_from_community",
+        }.issubset(df_imported_energy.columns):
+            raise ValueError(
+                "df_imported_energy must contain columns: simulation_time, \
+                imported_from_grid, imported_from_community"
+            )
+
+        df_carbon_ratio.index = pd.to_datetime(df_carbon_ratio.index)
+        df_imported_energy["simulation_time"] = pd.to_datetime(
+            df_imported_energy["simulation_time"], utc=True
+        )
+
+        # Find the nearest ratio for each df_imported_energy timestamp
+        df_imported_energy["carbon_ratio"] = df_imported_energy["simulation_time"].apply(
+            lambda ts: df_carbon_ratio.iloc[
+                df_carbon_ratio.index.get_indexer([ts], method="nearest")[0]
+            ]
+        )
+        df_imported_energy["carbon_generated_base_case_g"] = (
+            df_imported_energy["imported_from_community"]
+            + df_imported_energy["imported_from_grid"]
+        ) * df_imported_energy["carbon_ratio"]
+
+        df_imported_energy["carbon_generated_gsy_g"] = (
+            df_imported_energy["imported_from_grid"] * df_imported_energy["carbon_ratio"]
+        )
+
+        df_imported_energy["carbon_savings_g"] = (
+            df_imported_energy["carbon_generated_base_case_g"]
+            - df_imported_energy["carbon_generated_gsy_g"]
+        )
+        return df_imported_energy
+
+    def calculate_from_gsy_trade_profile(self, country_code: str, trade_profile: Dict) -> Dict:
+        """Calculate the carbon emissions from gsy-e trade profile,
+        usually from InfiniteBus accummulated sold_energy"""
+
+        if len(trade_profile.keys()) == 0:
+            raise ValueError("Invalid trade profile")
+
+        start_date, end_date = self._find_start_and_end_dates(
+            trade_profile.keys(), "%B %d %Y, %H:%M h"
```

Comment:
Why does the trade profile have a different time format than the imported_exported energy?
We should definitely look into this. Can you please leave a TODO comment in code? Thanks!